src/Project/Project.router.ts

```diff
@@ -126,6 +131,10 @@ export class ProjectRouter extends Router {
       throw new HTTPError('Invalid schema', validator.errors)
     }
 
+    if (!(await Project.canUpsert(id, user_id))) {
```

Comment:
Why are we using this instead of the `projectAuthorization` middleware?

Reply:
Because projectAuthorizations does the following:
`await Project.isOwnedBy(id, user_id)`

but when you're upserting the project, the first insert will never work!
That's why we do this check:

```ts
static async canUpsert(id: string, userId: string) {
  const [projectExists, isOwner] = await Promise.all([
    Project.exists(id),
    Project.isOwnedBy(id, userId)
  ])
  return !projectExists || isOwner
}
```

you can upsert if the project does not exist (first time) or if you're the owner. The same thing happens on the manifest upsert

Does that make sense?

Reply:
Totally :+1: